FuelLabs/fuel-indexer
# enhancement: add graphql enum support

cc @lostman 

### Description
- PR does a ton of things
  - Fixes type issues in the explorer example 
  - Refactors GraphQL schema codegen and table creation (it was spaghetti code)
  - Adds to concept of "virtual columns" on type IDs
    - Not all types need tables to be created 
    - Not only was this needed for enums but is a precursor for https://github.com/FuelLabs/fuel-indexer/issues/940
  - Cleans up how we're doing imports for the codgen and more important adds comments/explanations
  - Adds new types
    - We now support indexing enum types
      - Enum types have to be persisted into the DB as some type of `FtColumn` so I chose `String`
        - Users can convert this `String` back into its enum on their side if they wish 
  - Adds tests

### Testing steps
- [ ] Tests were added so CI should pass
- [ ] Start a local node `cargo run -p fuel-node --bin fuel-node`
- [ ] Start a local web api `cargo run -p web-api --bin web-api`
- [ ] Start the service using the test indexer 
```bash
cargo run --bin fuel-indexer -- run --manifest packages/fuel-indexer-tests/components/indices/fuel-indexer-test/fuel_indexer_test.yaml --run-migrations
```
- [ ] Trigger the enum event `curl -X POST http://localhost:8000/enum`
- [ ] Open the playground `http://localhost:29987/api/playground/fuel_indexer_test/index1`
- [ ] Query for the indexed enum 

```graphql
query { complexenumentity { id one }}
```
Should return the object with its enum as a String

```json
{
  "data": [
    {
      "id": 1,
      "one": "EnumEntity::One"
    }
  ]
}
```


### Changelog
- Add GraphQL enum support
- Refactor schema and table codegen
